File: ruby/resource_file.rb
require 'set'

class AndroidFile
	JAVA = "java"
	XML = "xml"
	PNG = "png"

	attr_accessor :path, :file_name

	def initialize(path, file_name)
		@path = path
		@file_name = file_name
	end


	def get_full_path()
		return File.join(@path, @file_name)
	end
end

class AndroidResource < AndroidFile
	DRAWABLE = "drawable"
	LAYOUT = "layout"
	ANIM = "anim"
	COLOR = "color"
	MENU = "menu"
	VALUES = "values"

	attr_accessor :resource_name, :used, :type

	def initialize(android_file, res_type)
		super(android_file.path, android_file.file_name)
		init(res_type)
	end

	def init(res_type)
		@type = res_type
		@folder_type = @path.split("/")[-1].split("-")[0]
		@resource_name = @file_name.split(".")[0]
		@used = false
	end

end

class ResourceGroup
	attr_accessor :resources

	def initialize()
		@resources = Hash.new
		@used_resources = Set.new
	end

	def insert(resource)
		if @resources.has_key?(resource.type) == false
			@resources[resource.type] = Array.new
		end
		@resources[resource.type].push(resource)
	end

	def set_used(resource_type, resource_name)
		if @resources.has_key?(resource_type)
			for resource in @resources[resource_type]
				if resource.resource_name == resource_name
					resource.used = true
				end
			end
		end
	end

	def get_contain_types()
		return @resources.keys()
	end

	def get_resources(resource_type)
		if @resources.has_key?(resource_type)
			return @resources[resource_type]
		else
			return Array.new
		end
	end

	def size()
		return @resources.length
	end

	def size(resource_type)
		if @resources.has_key?(resource_type)
			return @resources[resource_type].length
		else 
			return 0
		end
	end
end

class ResourceFinder
	# REGEX_PACKAGE_IN_MANIFEST = "\Wpackage\W+\""
	ANDROID_MANIFEST ="AndroidManifest.xml"

	def get_manifest_path()
		return File.join(@project_path, ANDROID_MANIFEST)
	end

	def get_package_in_manifest()
		#TODO find with regex
		return "com.vingle.android"
	end

	def initialize(project_path)
		@project_path = project_path
		@files = Hash.new
		@resources = ResourceGroup.new

		Dir.chdir(@project_path)
	end

	def find_base_env()
		@manifest = AndroidFile.new(@project_path, ANDROID_MANIFEST)
		@package = get_package_in_manifest()
	end

	def find_files_in(path, ext)
		file_filter = File.join(path, "**", "*.#{ext}")
		file_list = Array.new
		Dir.glob(file_filter) do |file|
			android_file = AndroidFile.new(File.dirname(file), File.basename(file))
			file_list.push(android_file)
		end
		return file_list
	end

	def find_files()
		@files[AndroidFile::JAVA] = find_files_in("src", AndroidFile::JAVA)
		@files[AndroidFile::XML] = find_files_in("res", AndroidFile::XML)
		@files[AndroidFile::PNG] = find_files_in("res", AndroidFile::PNG)
	end

	def find_exist_resources()
		@files[AndroidFile::PNG].each do |png_file|
			@resources.insert(AndroidResource.new(png_file, AndroidResource::DRAWABLE))
		end

		@files[AndroidFile::XML].each do |xml_file|
			type = xml_file.path.split("/")[-1].split("-")[0]
			unless type == AndroidResource::VALUES
				@resources.insert(AndroidResource.new(xml_file, type))
			end
		end
	end

	def find_used_resources()
		@files[AndroidFile::JAVA].each do |java_file|
			find_resources_in_java(java_file)
		end

		@files[AndroidFile::XML].each do |xml_file|
			find_resources_in_xml(xml_file)
		end
		find_resources_in_xml(@manifest)
	end

	def find_resources_in_java(java_file)
		regex_code = /#{@package}\.R\.\w+\.\w+/
		regex_find_code = /\W+#{regex_code}/
		regex_import = /import\W+#{@package}\.R/

		file = File.open(java_file.get_full_path)

		# find import R 
		file.each do |line|
			if line =~ regex_import
				regex_code = /R\.\w+\.\w+/
				regex_find_code = /\W+#{regex_code}/
				break
			end
		end

		file.rewind

		#find resource
		file.each do |line|
			if line =~ regex_find_code
				line.scan(regex_code).each do |resource_code|
					tokens = resource_code.split(".")
					resource_type = tokens[-2]
					resource_name = tokens[-1]
					@resources.set_used(resource_type, resource_name)
				end
			end
		end
	end

	def find_resources_in_xml(xml_file)
		regex_code = /\@\w+\/\w+/

		File.open(xml_file.get_full_path).each do |line|
			if line =~ regex_code
				line.scan(regex_code).each do |resource_code|
					tokens = resource_code.split("/")
					resource_type = tokens[-2][1..-1]
					resource_name = tokens[-1]
					@resources.set_used(resource_type, resource_name)
				end
			end
		end
	end

	def print_unused_resources()
		@resources.resources.each do |type, resources|
			unused_items = Set.new
			resources.each do |resource|
				if resource.used == false
					unused_items.add(resource.resource_name)
				end
			end
			if unused_items.length > 0
				unused_items.each do |name|
					puts name
				end
			end
		end
	end
end






finder = ResourceFinder.new("/Users/kai/vingle/vingle-android")
puts finder.get_manifest_path

finder.find_base_env
finder.find_files
finder.find_exist_resources
finder.find_used_resources
finder.print_unused_resources

File: python/resource_file.py
import os

class AndroidFile(object):
	JAVA = "java"
	XML = "xml"
	PNG = "png"

	def __init__(self, path, file_name):
		self.path = path
		self.file_name = file_name

	def get_full_path(self):
		return os.path.join(self.path, self.file_name)


class AndroidResource(AndroidFile):
	DRAWABLE = "drawable"
	LAYOUT = "layout"
	ANIM = "anim"
	COLOR = "color"
	MENU = "menu"
	VALUES = "values"

	def __init__(self, android_file, res_type):
		AndroidFile.__init__(self, android_file.path, android_file.file_name)
		self.init(res_type)

	def init(self, res_type):
		self.type = res_type
		self.folder_type = self.path.split("/")[-1].split("-")[0]
		self.resource_name = self.file_name.split(".")[0]
		self.used = False

	def set_used(self, used = True):
		self.used = used

class ResourceGroup(object):
	def __init__(self):
		self.resources = {}
		self.used_resources = set()
	def __repr__(self):
		ret_dict = {}
		for key in self.resources:
			ret_dict[key] = len(self.resources[key])
		return str(ret_dict)
	
	def insert(self, resource):
		if resource.type not in self.resources.keys():
			self.resources[resource.type] = []

		self.resources[resource.type].append(resource)

	def set_used(self, resource_type, resource_name):
		used_resource = (resource_type, resource_name);
		if used_resource not in self.used_resources:
			self.used_resources.add(used_resource)
			if resource_type in self.resources.keys():
				for resource in self.resources[resource_type]:
					if resource.resource_name == resource_name:
						resource.set_used()

	def get_contain_types(self):
		return self.resources.keys()

	def get_resources(self, resource_type):
		if resource_type not in self.resources.keys():
			return []
		return self.resources[resource_type]

	def size(self):
		return len(self.resources)

	def size(self, resource_type):
		return len(self.resources[resource_type])

